generate/generate_hcl_test.go

```diff
@@ -14,62 +14,339 @@
+					),
+				},
+				{
+					path: "/stacks/stack-2",
+					add: globals(
+						str("local_a", "stack-2-local"),
+						boolean("local_b", false),
+						number("local_c", 777),
+						attr("local_d", `{ oopsie = "local_d_field"}`),
+						str("backend_prefix", "stack-2-backend"),
+						str("provider_data", "stack-2-provider-data"),
+						str("provider_version", "stack-2-provider-version"),
+						str("terraform_version", "stack-2-terraform-version"),
+					),
+				},
+			},
+			want: []want{
+				{
+					stack: "/stacks/stack-1",
+					hcls: map[string]fmt.Stringer{
+						"backend.tf": backend(
+							labels("test"),
+							str("prefix", "stack-1-backend"),
+						),
+						"locals.tf": locals(
+							str("stackpath", "/stacks/stack-1"),
+							str("local_a", "stack-1-local"),
+							boolean("local_b", true),
+							number("local_c", 666),
+							str("local_d", "local_d_field"),
+						),
+						"provider.tf": hcldoc(
+							provider(
+								labels("name"),
+								str("data", "stack-1-provider-data"),
+							),
+							terraform(
+								required_providers(
+									expr("name", `{
+										source  = "integrations/name"
+										version = "stack-1-provider-version"
+									}`),
+								),
+							),
+							terraform(
+								str("required_version", "stack-1-terraform-version"),
+							),
+						),
+					},
+				},
+				{
+					stack: "/stacks/stack-2",
+					hcls: map[string]fmt.Stringer{
+						"backend.tf": backend(
+							labels("test"),
+							str("prefix", "stack-2-backend"),
+						),
+						"locals.tf": locals(
+							str("stackpath", "/stacks/stack-2"),
+							str("local_a", "stack-2-local"),
+							boolean("local_b", false),
+							number("local_c", 777),
+							attr("local_d", "null"),
+						),
+						"provider.tf": hcldoc(
+							provider(
+								labels("name"),
+								str("data", "stack-2-provider-data"),
+							),
+							terraform(
+								required_providers(
+									expr("name", `{
+										source  = "integrations/name"
+										version = "stack-2-provider-version"
+									}`),
+								),
+							),
+							terraform(
+								str("required_version", "stack-2-terraform-version"),
+							),
+						),
+					},
+				},
+			},
+		},
+	}
 
-func number(name string, val int64) hclwrite.BlockBuilder {
-	return hclwrite.NumberInt(name, val)
-}
+	for _, tcase := range tcases {
+		t.Run(tcase.name, func(t *testing.T) {
+			s := sandbox.New(t)
+			s.BuildTree(tcase.layout)
 
-func boolean(name string, val bool) hclwrite.BlockBuilder {
-	return hclwrite.Boolean(name, val)
-}
+			for _, cfg := range tcase.configs {
+				path := filepath.Join(s.RootDir(), cfg.path)
+				test.AppendFile(t, path, config.Filename, cfg.add.String())
+			}
 
-func labels(labels ...string) hclwrite.BlockBuilder {
-	return hclwrite.Labels(labels...)
-}
+			workingDir := filepath.Join(s.RootDir(), tcase.workingDir)
+			err := generate.Do(s.RootDir(), workingDir)
+			assert.IsError(t, err, tcase.wantErr)
```

Comment:
are we actually setting `wantErr` in any of the test cases? if this is not the case, shouldn't this assertion fail?

Reply:
If we never set wantErr, meaning that we are not expecting a failure, and no failure happens, this is the same as:

```go
assert.IsError(t, nil, nil)
```

Which works, because nil == nil (they match). If we get an unexpected error, it fails, and if we define wantErr then it would fail if the got err doesn't match.

Reply:
But since we are not using it here we can remove it, I usually have some error handling tests but so far haven't added at this level. But here you can see how it works:

https://github.com/mineiros-io/terramate/blob/ad148a17de74f6f7b6e4594c69188f920f265483/generate/genhcl/genhcl_test.go#L577

When you set the error, it will check for a specific match with the sentinel wrapped inside the error, when you don't set it it will check if both errors are nil, since only a nil error "is" another nil error.

Reply:
Simpler example: https://go.dev/play/p/aMat5TWAU7t

Reply:
got it! thanks for taking the time to explain! :heart: